### ISSUE TITLE
Find-Files in nugetinstaller task fails with path name to long

### ISSUE THREAD

**Issue by ro-** [Microsoft/vso-agent-tasks#1111]
I noticed once i started using npm packages in my project, if the node_modules folder exists when Find-Files command is executed in the nuget package restore it will fail with a path too long exception.

Recursive node_modules is not really a problem to solve here, but a work around where a folder can be ignored would be good?

For now im restoring nuget packages, then running npm install, then building the solution. However I need to have git source clean turned on (which slows things down a lot) otherwise the node_modules exists from the last build.

```
2016-01-22T04:27:10.1109667Z Set workingFolder to default: C:\Program Files\Microsoft Team Foundation Server 14.0\Build\tasks\NuGetInstaller\0.1.12
 
2016-01-22T04:27:10.1266212Z Executing the powershell script: C:\Program Files\Microsoft Team Foundation Server 14.0\Build\tasks\NuGetInstaller\0.1.12\NuGetInstaller.ps1
 
2016-01-22T04:27:10.2205450Z Entering script System.Management.Automation.InvocationInfo.MyCommand.Name
 
2016-01-22T04:27:10.2205450Z Parameter Values
 
2016-01-22T04:27:10.2205450Z solution = C:\T\f6b74389\Productionwise\Productionwise.sln
 
2016-01-22T04:27:10.2362485Z noCache = false
 
2016-01-22T04:27:10.2362485Z nuGetRestoreArgs = -verbose
 
2016-01-22T04:27:10.2362485Z nuGetPath = 
 
2016-01-22T04:27:10.2518732Z No Pattern found in solution parameter.
 
2016-01-22T04:27:10.2518732Z Searching for nuget package configuration files using pattern C:\T\f6b74389\Productionwise\**\packages.config
 
2016-01-22T04:27:11.0187525Z ##[error]The specified path, file name, or both are too long. The fully qualified file name must be less than 260 characters, and the directory name must be less than 248 characters.
 
2016-01-22T04:27:11.0187525Z No nuget package configuration files found for C:\T\f6b74389\Productionwise\Productionwise.sln
```

**Comment by ericsciple** [comment]
Yes, unfortunately it's currently a problem. It's been fixed in the new task lib that is currently being finalized and a couple tasks have been ported over to use it.

One possible workaround might be for you to narrow your search path to avoid the node modules folder being traversed. Another workaround is to give the literal path to the file you want to run restore against - if the path contains wildcards, then the task calls Find-Files. Another workaround is a hybrid of the first two options - multiple nuget restore tasks and literal paths/narrower search paths.

**Comment by ro-** [comment]
Yes avoiding node_modules would be good, though pulling apart the powershell script in the task, it seems that theres no way to avoid it running that line with the built in task. 

Ultimately the Find-files command is the problem. I get the same error on a test assemblies task later on. 

I can't seem to locate any documentation for that command either.

**Comment by ymerej** [comment]
I have the same problem.  I used ILSpy to look at the code within the Microsoft.TeamFoundation.DistributedTask.Task.Common.dll assembly.  I am not finding a git repo for this assembly.  I'd be willing to submit a pull request for this bug if I could find the repo.

There are two methods involved in the bug.  The GetMatchingItems method is responsible for excluding paths returned by the GetPathIterator method.  The problem is that GetPathIterator does a full depth search before returning the results.  An exception occurs before the results of GetPathIterator can be returned to GetMatchingItems for exclusion.

**Comment by ericsciple** [comment]
It's supported in the new lib (https://github.com/Microsoft/vsts-task-lib/tree/master/powershell/VstsTaskSdk). Assigning the issue to Matt.

**Comment by vtbassmatt** [comment]
Hey @ericsciple not sure what's expected of us here. Do existing tasks not inherit the new Find-Files behavior?

**Comment by ericsciple** [comment]
@vtbassmatt No, the existing task handler has a lot of problems that can't be solved by patching it. We've stepped back and designed a new handler to address pain-points of the existing handler. For example some of the pain points with the old handler are: runs in proc in the worker process, issues w/ stderr, can't load the lib and interactively test without actually running via the agent, lib is not open sourced, tight coupling with the agent version which lead to version mismatch hurdles every time we want to change functionality within the compiled cmdlets, there's also the issue of maintaining back compat in the compiled cmdlet (the task authors chooses the lib version in the new handler).

We've added a new open source lib that is designed to work with the new handler. Docs for the lib are here: https://github.com/Microsoft/vsts-task-lib/blob/master/powershell/Docs/VstsTaskSdk.md The documentation still needs to be added wrt each individual command. There's still some good info in the docs right now though. As for examples, three tasks have been ported over MSBuild/VSBuild/PublishSymbols. In 94, consuming the new tasks is behind a flag (set build variable system.preferps3 to true). In 95 we're flipping it so that system.preferpsoverps3 needs to be set to true to revert back to the old task. Probably in 96 we can delete the legacy implementations of msbuild/vsbuild/publishsymbols and leave the flag there for anyone who needs to port their tasks.

**Comment by ericsciple** [comment]
@vtbassmatt On a somewhat related note, there's also a mechanism now in the gulpfile to handle (internal) common code that is shared across tasks. I remember you guys have shared code between the NuGet tasks. For an example, you can see how the msbuild and vsbuild tasks reference shared code (from the Common folder). It just takes an entry in the common.json in the repo root. Build process handles copying the dependency in place.

**Comment by vtbassmatt** [comment]
Thanks @ericsciple. We'll do the reaction work in S96.
